docs/development/design/embed-api.rst

```diff
@@ -0,0 +1,211 @@
+               "id": "check-that-everything-works",
+               "url": "https://docs.readthedocs.io/en/latest/development/install.html#check-that-everything-works"
+            },
+            ...
+         ]
+      }
+
+
+Handle specific Sphinx cases
+----------------------------
+
+.. https://github.com/readthedocs/readthedocs.org/pull/8039#discussion_r640670085
+
+We are currently handling some special cases for Sphinx due how it writes the HTML output structure.
+In some cases, we look for the HTML tag with the identifier requested but we return
+the ``.next()`` HTML tag or the ``.parent()`` tag instead of the *requested one*.
+
+Currently, we have identified that this happens for definition tags (``dl``, ``dt``, ``dd``)
+--but may be other cases we don't know yet.
+Sphinx adds the ``id=`` attribute to the ``dt`` tag, which contains only the title of the definition,
+but as a user, we are expecting the description of it.
+
+In the following example we will return the whole ``dl`` HTML tag instead of
+the HTML tag with the identifier ``id="term-name"`` as requested by the client,
+because otherwise the "Term definition for Term Name" content won't be included and the response would be useless.
+
+.. code:: html
+
+   <dl class="glossary docutils">
+     <dt id="term-name">Term Name</dt>
+     <dd>Term definition for Term Name</dd>
+   </dl>
+
+If the definition list (``dl``) has more than *one definition* it will return **only the term requested**.
+Considering the following example, with the request ``?url=glossary.html#term-name``
+
+.. code:: html
+
+   <dl class="glossary docutils">
+     ...
+
+     <dt id="term-name">Term Name</dt>
+     <dd>Term definition for Term Name</dd>
+
+     <dt id="term-unknown">Term Unknown</dt>
+     <dd>Term definition for Term Unknown </dd>
+
+     ...
+   </dl>
+
+
+It will return the whole ``dl`` with only the ``dt`` and ``dd`` for ``id`` requested:
+
+.. code:: html
+
+   <dl class="glossary docutils">
+     <dt id="term-name">Term Name</dt>
+     <dd>Term definition for Term Name</dd>
+   </dl>
+
+
+However, this assumptions may not apply to documentation pages built with a different doctool than Sphinx.
+For this reason, we need to communicate to the API that we want to handle this special cases in the backend.
+This will be done by appending a request GET argument to the Embed API endpoint: ``?doctool=sphinx&version=4.0.1``.
+In this case, the backend will known that has to deal with these special cases.
+
+.. note::
+
+   This leaves the door open to be able to support more special cases (e.g. for other doctools) without breaking the actual behavior.
+
+
+Support for external documents
+------------------------------
+
+When the ``?url=`` argument passed belongs to a documentation page not hosted on Read the Docs,
+the endpoint will do an external request to download the HTML file,
+parse it and return the content for the identifier requested.
```

Comment:
I'm a bit hesitant on this concept. Requesting remote content does open up some security concerns. It seems like this is primarily just to get around CORS headers? Otherwise, this seems like it would be better implemented as a client side function, not an RTD hosted service.

To add to the security concerns you outlined, anything purposely or accidentally tying up our web processes, like latent servers and large file requests, and proxying abusive traffic through our server requests to mask origins.

Reply:
>  It seems like this is primarily just to get around CORS headers?

Nope, is required for the grant to make the api work with sites hosted outside RTD. I'm also worried about supporting external sites, don't know if we already committed to have that feature or not /cc @ericholscher

Reply:
There are some security concerns we should manage very carefully, for sure.

The main goal of this feature is to be able to "embed the whole Internet" 😄, not to get around CORS. As an example, currently, if you are using `sphinx-hoverxref` together with intersphinx for external projects and the official Python documentation is one of those projects, you have to disable intersphinx tooltips _only_ for the Python docs because they are not hosted on RTD and tooltips doesn't work. By supporting embedding content from external sites, we could support this case.

Reply:
We haven't committed to this for the grant. It would likely start out as an allow list, not allowing any random site on the internet. The main goal here is CDN caching to improve the UX, and making it work across the docs ecosystem, not just things hosted on RTD.